response.go

```diff
@@ -2,23 +2,88 @@ package tarantool
 
 import (
 	"fmt"
+	"io"
+	"io/ioutil"
 
 	"github.com/tarantool/go-iproto"
 	"github.com/vmihailenco/msgpack/v5"
 )
 
-type Response struct {
-	RequestId uint32
-	Code      uint32
-	// Error contains an error message.
-	Error string
-	// Data contains deserialized data for untyped requests.
-	Data []interface{}
-	// Pos contains a position descriptor of last selected tuple.
-	Pos      []byte
-	MetaData []ColumnMetaData
-	SQLInfo  SQLInfo
-	buf      smallBuf
+type Response interface {
+	// Header returns a response header.
+	Header() Header
+	// Decode decodes a response.
+	// After a first Decode call,
+	// all next calls will return the result of a first one
+	// and no decoding will be done.
+	Decode() ([]interface{}, error)
+	// DecodeTyped decodes a response into a given container res.
+	// This method could be called multiple times.
+	DecodeTyped(res interface{}) error
+}
+
+type baseResponse struct {
+	// header is a response header.
+	header Header
+	// data contains deserialized data for untyped requests.
+	data []interface{}
+	buf  smallBuf
+	// Was the Decode() func called for this response.
+	decoded bool
+	// Was the DecodeTyped() func called for this response.
+	decodedTyped bool
+	err          error
+}
+
+func createBaseResponse(header Header, body io.Reader) (baseResponse, error) {
+	if body == nil {
+		return baseResponse{header: header}, nil
+	}
+	if buf, ok := body.(*smallBuf); ok {
+		return baseResponse{header: header, buf: *buf}, nil
+	}
+	data, err := ioutil.ReadAll(body)
+	if err != nil {
+		return baseResponse{}, err
+	}
+	return baseResponse{header: header, buf: smallBuf{b: data}}, nil
+}
+
+// PushResponse is used for push requests for the Future.
+type PushResponse struct {
+	baseResponse
+}
+
+func createPushResponse(header Header, body io.Reader) (Response, error) {
+	resp, err := createBaseResponse(header, body)
+	if err != nil {
+		return nil, err
+	}
+	return &PushResponse{resp}, nil
+}
+
+// SelectResponse is used for the select requests.
+// It might contain a position descriptor of the last selected tuple.
+type SelectResponse struct {
+	baseResponse
+	// pos contains a position descriptor of last selected tuple.
+	pos []byte
+}
+
+// PrepareResponse is used for the prepare requests.
+// It might contain meta-data and sql info.
+type PrepareResponse struct {
+	baseResponse
+	metaData []ColumnMetaData
+	sqlInfo  SQLInfo
+}
```

Comment:
It seems that it equal at now, but something may change in the future. So let's just make an alias.
```suggestion
alias PrepareResponse ExecuteResponse
```